### PR TITLE
1557-develop-jurong-island-input-agent: Added input agent folder unde…

### PR DESCRIPTION
**Pull request by nagarajankarthik**
Added Jurong Island input agent.

### CODE REVIEW

File: JPS_VIRTUALSENSOR/AermodAgent/src/main/java/com/cmclinnovations/aermod/Buildings.java [hunk 1]
```diff
@@ -1208,7 +1208,7 @@ public void getProperties() throws org.apache.jena.sparql.lang.sparql_11.ParseEx
         // Remove IRIs of pollutant sources
         for (int i = 0; i < BuildingIRIQueryResult.length(); i++){
             String buildingIRI = BuildingIRIQueryResult.getJSONObject(i).getString("cityObject");
-            if (!StackIRIString.contains(buildingIRI.replace("cityobject", "building"))) {
+            if (!StackIRIString.contains(buildingIRI.replace("cityobject", "cityfurniture"))) {
```
Comment (kok-foong) [inline JPS_VIRTUALSENSOR/AermodAgent/src/main/java/com/cmclinnovations/aermod/Buildings.java:1211]: is this change intentional in this branch?

File: JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/QueryClient.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    private static final Prefix P_BUILD = SparqlBuilder.prefix("build",iri(ONTO_BUILD));
+
+    // classes
+    // Since there is only CO2 emissions data for Jurong Island, other pollutants like NO2 and PM are 
+    // not listed here. 
+    private static final Iri StaticPointSource = P_DISP.iri("StaticPointSource");
+    private static final Iri CO2 = P_DISP.iri("CO2");
+    private static final Iri Density = P_OM.iri("Density");
+    private static final Iri MassFlow = P_OM.iri("MassFlow");
+    private static final Iri Temperature = P_OM.iri("Temperature");
```
Comment (kok-foong) [inline JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/QueryClient.java:73]: use screaming snake case for constants

File: JPS_VIRTUALSENSOR/JurongIslandInputAgent/pom.xml [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+		<!-- Testing, versions pulled from parent -->
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+        </dependency>
```
Comment (kok-foong) [inline JPS_VIRTUALSENSOR/JurongIslandInputAgent/pom.xml:101]: please also go through this and remove dependencies that are not needed by this agent

File: JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/ValuesPattern.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package com.cmclinnovations.jurongisland;
+import java.util.List;
+
+import org.eclipse.rdf4j.sparqlbuilder.core.Variable;
+import org.eclipse.rdf4j.sparqlbuilder.graphpattern.GraphPattern;
+import org.eclipse.rdf4j.sparqlbuilder.rdf.Iri;
+
+public class ValuesPattern<T> implements GraphPattern {
```
Comment (kok-foong) [inline JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/ValuesPattern.java:8]: I don't think this class is needed?

File: JPS_VIRTUALSENSOR/AermodAgent/src/main/java/com/cmclinnovations/aermod/Buildings.java (hunk 1 above)
Comment (nagarajankarthik) [reply JPS_VIRTUALSENSOR/AermodAgent/src/main/java/com/cmclinnovations/aermod/Buildings.java:1211]: Yes that change is intentional. For JurongIsland, we use the cityfurniture graph to get the height and diameter of pollutant sources. After searching for buildings in a bounding box, we get IRIs from the cityobject graph. To remove the pollutant emitting points from the results list, we need to replace cityobject with cityfurniture when checking if the IRI already exists in the list of pollutant emitting points.

Comment (kok-foong) [reply JPS_VIRTUALSENSOR/AermodAgent/src/main/java/com/cmclinnovations/aermod/Buildings.java:1211]: If I understand this correctly, this change needs to be committed together with following changes:
1) city gml namespace as input
2) aermod agent querying emission sources directly

Hence I don't think this should be in this branch because it interferes with the functionality of this branch, correct me if I'm wrong

Comment (nagarajankarthik) [reply JPS_VIRTUALSENSOR/AermodAgent/src/main/java/com/cmclinnovations/aermod/Buildings.java:1211]: I wouldn't say it interferes with the functionality of this branch since all the new functionality is within the JurongIslandInputAgent folder.

File: JPS_VIRTUALSENSOR/JurongIslandInputAgent/README.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+Follow the instructions in Deploy/stacks/dynamic/stack-manager/README.md to spin up a stack.
+
+# API
+Start the agent by executing the following command from within this folder. Replace STACK_NAME with the actual stack name:
+
+./stack.sh start <STACK_NAME>
+
+Send a GET reuest to the agent as follows from an Ubuntu terminal:
+
+curl -X GET http://localhost:8082/JurongIslandInputAgent/update
```
[the remark this answers is not in the document]
Comment (nagarajankarthik) [reply JPS_VIRTUALSENSOR/JurongIslandInputAgent/README.md]: Done

File: JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/ValuesPattern.java (hunk 4 above)
Comment (nagarajankarthik) [reply JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/ValuesPattern.java]: That class has been removed.

File: JPS_VIRTUALSENSOR/JurongIslandInputAgent/pom.xml (hunk 3 above)
Comment (nagarajankarthik) [reply JPS_VIRTUALSENSOR/JurongIslandInputAgent/pom.xml]: The unnecessary junit dependencies have been removed.

File: JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/QueryClient.java (hunk 2 above)
Comment (nagarajankarthik) [reply JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/QueryClient.java]: The case of those constants has been updated

File: JPS_VIRTUALSENSOR/AermodAgent/src/main/java/com/cmclinnovations/aermod/Buildings.java (hunk 1 above)
Comment (kok-foong) [reply JPS_VIRTUALSENSOR/AermodAgent/src/main/java/com/cmclinnovations/aermod/Buildings.java:1211]: Are you sure? I mean if you merge this into the 1461-create-an-aermod-agent-that-retrieves-building-related-information-from-the-knowledge-graph branch but still query the emissions using the old way won't the results change?

File: JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/EndpointConfig.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	private String kguser;
+	private String kgpassword;
+    
+    public EndpointConfig() {
+        ContainerClient containerClient = new ContainerClient();
+        PostGISEndpointConfig postGISEndpointConfig = containerClient.readEndpointConfig("postgis",
+                    PostGISEndpointConfig.class);
+        this.dburl = postGISEndpointConfig.getJdbcURL(EnvConfig.DATABASE);
+        this.dbuser = postGISEndpointConfig.getUsername();
+        this.dbpassword = postGISEndpointConfig.getPassword();
```
Comment (kok-foong) [inline JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/EndpointConfig.java:22]: I believe these can be removed

File: JPS_VIRTUALSENSOR/JurongIslandInputAgent/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  jurong-island-input-agent:
+    image: ghcr.io/cambridge-cares/jurong-island-input-agent:1.0.0-1557-develop-jurong-island-input-agent-SNAPSHOT
+    deploy:
+      mode: replicated
+      replicas: 1
+      restart_policy:
+        condition: none
+    environment:
+      - STACK_NAME=${STACK_NAME}
+      - DATABASE=postgres
```
Comment (kok-foong) [inline JPS_VIRTUALSENSOR/JurongIslandInputAgent/docker-compose.yml:13]: not needed?

File: JPS_VIRTUALSENSOR/AermodAgent/src/main/java/com/cmclinnovations/aermod/Buildings.java (hunk 1 above)
Comment (nagarajankarthik) [reply JPS_VIRTUALSENSOR/AermodAgent/src/main/java/com/cmclinnovations/aermod/Buildings.java:1211]: Actually, the query has been updated  in the 1461-create-an-aermod-agent-that-retrieves-building-related-information-from-the-knowledge-graph branch itself. Previously, I queried the OCGML IRIs of buildings from the same namespace that contains the data for emissions and OCGML IRIs of pollutant sources. I removed the call to the method in BuildingsQueryClient that executes that query. Instead, I now use the search within the bounding box to get buildings near pollutant emitting points.

Comment (kok-foong) [reply JPS_VIRTUALSENSOR/AermodAgent/src/main/java/com/cmclinnovations/aermod/Buildings.java:1211]: if that's the case make sure to increment the version number before merging

File: JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/EndpointConfig.java (hunk 6 above)
Comment (nagarajankarthik) [reply JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/EndpointConfig.java:22]: Oh ok I thought EndPointConfig.java is a standard file used by many agents so I didn't change it

**Review by kok-foong — APPROVED** [review]

File: JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/EndpointConfig.java (hunk 6 above)
Comment (nagarajankarthik) [reply JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/EndpointConfig.java]: EndPointConfig.java has been updated to remove the POSTGIS part

File: JPS_VIRTUALSENSOR/JurongIslandInputAgent/docker-compose.yml (hunk 7 above)
Comment (nagarajankarthik) [reply JPS_VIRTUALSENSOR/JurongIslandInputAgent/docker-compose.yml]: The STACK_NAME and DATABASE environment variables have been removed.

Comment (kok-foong) [reply JPS_VIRTUALSENSOR/JurongIslandInputAgent/docker-compose.yml]: STACK_NAME is required I believe? Sorry i was not clear, was just referring to DATABASE

File: JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/QueryClient.java [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        List<String> ocgmlIRIs = Arrays.asList(
+            "http://www.theworldavatar.com:83/citieskg/namespace/pirmasensEPSG32633/sparql/cityobject/UUID_LOD2_Pirmasens_4f8d0f1a-3b21-40d4-8b90-89723e31a7ca/",
+            "http://www.theworldavatar.com:83/citieskg/namespace/pirmasensEPSG32633/sparql/cityobject/UUID_LOD2_Pirmasens_c38d038b-a677-4e0c-95d9-f02c09cf991c/");
+            List<Double> no2Emissions = Arrays.asList(100.0,100.0);
+            List<Double> pm25Emissions = Arrays.asList(100.0,100.0);
+            List<Double> pm10Emissions = Arrays.asList(100.0,100.0);
+
+            List<Double> no2Densities = Arrays.asList(1.0,1.0);
+            List<Double> pm25Densities = Arrays.asList(1.0,1.0);
+            List<Double> pm10Densities = Arrays.asList(1.0,1.0);
```
Comment (kok-foong) [inline JPS_VIRTUALSENSOR/JurongIslandInputAgent/src/main/java/com/cmclinnovations/jurongisland/QueryClient.java:150]: In the future, try to use object oriented programming instead of creating 7 lists
